image_eval/evaluators.py

```diff
@@ -1,100 +1,135 @@
 import abc
-from typing import Optional
 
 import ImageReward as RM
 import clip
 import torch
 from PIL import Image
+from enum import Enum
 from torchmetrics.image.fid import FrechetInceptionDistance
 from torchmetrics.image.inception import InceptionScore
 from torchmetrics.multimodal.clip_score import CLIPScore
 from torchvision import transforms
+from typing import Dict
 from vendi_score import vendi
 
 from image_eval.improved_aesthetic_predictor import run_inference
 from image_eval.encoders import ALL_ENCODER_CLASSES
+from image_eval.model_utils import download_model
 
 torch.manual_seed(42)
 
 
 # TODO (mihail): Decouple this so not all evaluators are in the same file
 
-class BaseReferenceFreeEvaluator(abc.ABC):
-    """
-    An evaluation that doesn't require gold samples to compare against.
-    """
-    def __init__(self, device: str, model_path: Optional[str] = None):
-        self.device = device
-        self.model_path = model_path
 
-    @abc.abstractmethod
-    def evaluate(self, images: list[Image.Image], prompts: list[str]):
-        pass
+class EvaluatorType(Enum):
+    """Follows the evaluation terminology established by https://arxiv.org/pdf/2309.14859.pdf."""
+    # The visual appeal of the generated images (naturalness, absence of artifacts or deformations).
+    IMAGE_QUALITY = 1
 
+    # The model’s ability to generate images that align well with text prompts.
+    CONTROLLABILITY = 2
 
-class BaseWithReferenceEvaluator(abc.ABC):
-    """
-    An evaluation that includes gold samples to compare against.
-    """
-    def __init__(self, device: str, model_path: Optional[str] = None):
+    # The extent to which generated images adhere to the target concept. Relevant for fine-tuned
+    # models rather than foundational ones (e.g. vanilla Stable Diffusion).
+    FIDELITY = 3
+
+    # The variety of images that are produced from a single or a set of prompts.
+    DIVERSITY = 4
+
+
+class BaseEvaluator(abc.ABC):
+    """Base class for evaluators."""
+    def __init__(self, device: str):
         self.device = device
-        self.model_path = model_path
 
     @abc.abstractmethod
-    def evaluate(self, generated_images: list[Image.Image], real_images: list[Image.Image]):
+    def evaluate(self, generated_images: list[Image.Image], *args, **kwargs) -> Dict[str, float]:
         pass
 
+    def is_useful(self, generated_images: list[Image.Image], *args, **kwargs) -> bool:
```

Comment:
please change the name to be more descriptive (i.e. `enough_data_for_evaluator` or something)

Reply:
Updated to `should_trigger_for_data`